WooCommerce/Classes/ViewModels/Order Details/Shipping Labels/ShippingLabelOrderItemsAggregator.swift

```diff
@@ -0,0 +1,153 @@
+    init(shippingLabels: [ShippingLabel],
+         orderItems: [OrderItem],
+         products: [Product],
+         productVariations: [ProductVariation],
+         currencyFormatter: CurrencyFormatter = CurrencyFormatter(currencySettings: ServiceLocator.currencySettings)) {
+        self.currencyFormatter = currencyFormatter
+
+        aggregateProductsToOrderItems(shippingLabels: shippingLabels, orderItems: orderItems, products: products, productVariations: productVariations)
+    }
+
+    /// Returns an array of order items for a shipping label.
+    func orderItems(of shippingLabel: ShippingLabel) -> [AggregateOrderItem] {
+        orderItemsByShippingLabelID[shippingLabel.shippingLabelID] ?? []
+    }
+}
+
+private extension ShippingLabelOrderItemsAggregator {
+    mutating func aggregateProductsToOrderItems(shippingLabels: [ShippingLabel],
+                                                orderItems: [OrderItem],
+                                                products: [Product],
+                                                productVariations: [ProductVariation]) {
+        shippingLabels.forEach { aggregateProductsToOrderItems(shippingLabel: $0,
+                                                               orderItems: orderItems,
+                                                               products: products,
+                                                               productVariations: productVariations) }
+    }
+
+    mutating func aggregateProductsToOrderItems(shippingLabel: ShippingLabel,
+                                                orderItems: [OrderItem],
+                                                products: [Product],
+                                                productVariations: [ProductVariation]) {
+        // ShippingLabel's `productNames` is always available, but `productIDs` is only available in WooCommerce Shipping & Tax v1.24.1+.
+        // Here we map a ShippingLabel's `productNames` to `ProductInformation` with an optional product ID at the corresponding index.
+        let productInfoArray = shippingLabel.productNames.enumerated().map { index, productName in
+            ProductInformation(id: shippingLabel.productIDs[safe: index], name: productName)
+        }
+        // In the API, if a product has quantity higher than 1, it is repeated in the `productNames`/`productIDs` array.
+        // Note that decimal quantity might have unexpected behavior given the current API design.
+        let uniqueProductInfoArray = productInfoArray.removingDuplicates()
+
+        // Generates a dictionary that maps a `ProductInformation` to the number of times (quantity) in a shipping label.
+        let countsByProductInformation = productInfoArray.reduce([:]) { (result, productInfo) -> [ProductInformation: Int] in
+            var countsByProductInfo = result
+            countsByProductInfo[productInfo] = (result[productInfo] ?? 0) + 1
+            return countsByProductInfo
+        }
```

Comment:
ℹ️ `reduce(into:)` is more succinct and less awkward. 🙂 

```suggestion
        let countsByProductInformation = productInfoArray.reduce(into: [ProductInformation: Int]()) { result, productInfo in
            result[productInfo] = (result[productInfo] ?? 0) + 1
        }
```

But up to you!

Reply:
wow TIL'ed `reduce(into:)`, this looks much more concise! updated in https://github.com/woocommerce/woocommerce-ios/pull/3197/commits/7dbe1f3ebf33f3e571285748bc76007225b1618f